### ISSUE TITLE
[BUG] Problem with excluding files from analysis in VSCode (only for Windows)

### ISSUE THREAD

**Issue by roman-petrov** [dart-code-checker/dart-code-metrics#268]
I am trying to set up `dart_code_metrics` and Visual Studio Code. Everything works fine from command line but not in IDE.

Here is a very simple project I created to reproduce the issue:
[test_app.zip](https://github.com/dart-code-checker/dart-code-metrics/files/6317430/test_app.zip)

Note that `src/exclude_me.dart` is excluded from analysis and `prefer_single_quotes` analysis error is not shown in Visual Studio Code. But I still get `newline-before-return` analysis error from `dart_code_metrics`. 

I tried both approaches to exclude files: `exclude` option for `analyzer` and `metrics-exclude` option for `dart_code_metrics` - both give the same result. Everything is ok from command line but not in IDE.

I am not sure where is the source of this issue: `dart_code_metrics` plugin, `Dart/Flutter` VSCode extension or even `analyzer` package / analysis server.  I am not sure if I am doing something wrong, can you please assist?

**Comment by incendial** [comment]
Hi, I've downloaded the example and everything works as expected (no highlights in VSCode). I'm kinda suspicious about plugin manager cache, but let's check preconditions first:
1. Do you have LSP preview enabled?
2. Could you please check the 5th step [from the doc](https://github.com/dart-code-checker/dart-code-metrics/blob/master/TROUBLESHOOTING.md) and share the screenshot (pay additional attention to amount of plugins activated)?

**Comment by roman-petrov** [comment]
Thank you for quick response. It's actually quite good news that everything works fine on your environment:) So there are the information you asked for:

1. I am not using LSP in this test application: my user settings in VSCode are almost empty. In LSP mode (`"dart.previewLsp": true`) results are almost the same. 
>LSP has another issue - line number in issue is reported incorrectly, but this might be `Dart Code` extension problem. 
2. Here is a screenshot for "Plugins" in analyzer diagnostics:
![image](https://user-images.githubusercontent.com/18419515/114968827-d23a3f80-9e7f-11eb-90d2-5eb11779a827.png)

Finally, here is a screenshot from my VSCode with the issue:

![image](https://user-images.githubusercontent.com/18419515/114969232-a66b8980-9e80-11eb-9fdf-9c947f6561f1.png)

So, when we figured out that this is environment related issue I currently see some possible reasons / steps to solve:

1. I will ask colleagues to try to reproduce the problem in the test application on their machines 
2. I will try to clear analyzer cache. Are these path from analyzer diagnostics correct?
```
bootstrap package path: C:\Users\Roman Petrov\AppData\Local\Pub\Cache\hosted\pub.dartlang.org\dart_code_metrics-3.0.0\tools\analyzer_plugin

execution path: C:\Users\Roman Petrov\AppData\Local\.dartServer\.plugin_manager\3a198a292c388ca5643c46b018d48f42\analyzer_plugin\bin\plugin.dart

packages file path: C:\Users\Roman Petrov\AppData\Local\.dartServer\.plugin_manager\3a198a292c388ca5643c46b018d48f42\analyzer_plugin\.packages
```
3. We use master Flutter SDK channel so I can try to switch to stable
Flutter: `Channel unknown, 2.2.0-11.0.pre.51, on Microsoft Windows [Version 10.0.19041.867], locale en-US`
Dart: `Dart SDK version: 2.13.0-219.0.dev (dev) (Thu Apr 8 11:44:31 2021 -0700) on "windows_x64"`
4. I will try to disable all extensions in VSCode except `Dart / Flutter`

Is there something else you can suggest to try?...

Thank you for all your help!

**Comment by roman-petrov** [comment]
UPD: I tried steps 2,3,4 but none of them helped me out.

1. I deleted folder `C:\Users\Roman Petrov\AppData\Local\.dartServer\.plugin_manager`
2. I switched to Flutter stable running `flutter channel stable` (`flutter doctor` now reports `2.0.4` installed)
3. I disabled all extensions in VSCode except Dart / Flutter

So I will try to ask my colleagues to check if they can reproduce the issue on their machines.

**Comment by incendial** [comment]
> I will try to clear analyzer cache. Are these path from analyzer diagnostics correct?

Yeah and I hoped that clearing will help...
If the issue is excluded correctly, it won't be even send to the analysis server, so I suppose other extensions or different SDK should not matter.

> Is there something else you can suggest to try?...

Two things come to mind:
1. replace the ignore pattern to
```yaml
exclude:
    - '**'
```
2. or add an inline ignore, like `//ignore_for_file: newline-before-return`

> Thank you for all your help!

Thank you for your feedback and providing great debug info, it's very important!

**Comment by roman-petrov** [comment]
1. I tried excluding `**` pattern in analysis options - this takes no effect. It seems `exclude` / `metrics-exclude` are completely ignored on my machine in VSCode.
2. `// ignore_for_file: newline-before-return` works just fine. BTW, there is no lightbulb in VSCode to create such autofix: I am not sure if it should appear for `dart_code_metrics`.

I will continue my research, maybe I will try to create an empty virtual machine to try a completely clean environment. Anyway, I will post here results of my investigations.

Thanks again for a great support and great contribution to Dart code quality!

**Comment by roman-petrov** [comment]
I am trying to explore analyzer logs from Dart Code extension, I [attached it](https://github.com/dart-code-checker/dart-code-metrics/files/6323203/analyzerLogFile1.txt) FYI.

There is some error shown at the end like `{"id":"22","error":{"code":"GET_FIXES_INVALID_FILE","message":"Error during 'edit.getFixes': invalid file."}}`. 

Not sure if this error might be source of my issue though. However, there is possible related but already closed issue in Dart SDK https://github.com/dart-lang/sdk/issues/43521

**Comment by incendial** [comment]
> UPD: This is getting quite interesting: I noticed that if I exclude file exclude_me.dart in analysis options - there is an error and no lightbulb with autofix is shown is VSCode.
> 
> But if exclude_me.dart file is not excluded in analysis options - lightbulb is shown in VSCode. Unfortunately I don't have good experience in Dart analyzer / analysis server / Dart Code extension to make any conclusions from this information.

Okay, that's definitely a bug from our side, I will fix it, thank you.
Have you succeeded with asking your colleagues? Maybe it's a Windows-related problem and that's why I can't reproduce it...
And sorry for long responses.

**Comment by roman-petrov** [comment]
I asked my colleague and he got the same result, he is able to reproduce the issue (he and almost everyone in our company uses Windows) . 

It's quite possible that this is a Windows related problem. Yesterday I started setting up a virgin Windows virtual machine :)
I will continue informing you when I get any updates.

Please let me know if you like to set up a call with screen sharing, I will be happy to provide any information that might help.

Actually, your responses are very short for me:) I'm quite excited about how great your support is!

**Comment by incendial** [comment]
@roman-petrov I've asked @dkrutskikh to check the problem on Windows machine and he confirmed that this is a bug with path building. Sorry for that long debug journey you have to make, I think we will fix it in the next minor release. I will also convert this issue to a bug type.

**Comment by roman-petrov** [comment]
This is great news! I will look forward to 3.1.0 release.

We plan to start using `dart_code_metrics` in our company to improve code quality in our products so I hope I will contribute to this great package not only with issues :)

Thank you for all your help!

**Comment by Alex-A4** [comment]
Same problem on MacOS Catalina 10.15.7.
Just add next lines, tried different exclude options like  - do not work.
```
analyzer:
  plugins:
    - dart_code_metrics

dart_code_metrics:
  rules:
    - prefer-trailing-comma
  exclude:
    - '**'
```

Also the problem is that analyzer tries to alanyze all the packages added to the project

**Comment by dkrutskikh** [comment]
@Alex-A4 your config has one mistake. Exclude section must be placed in `analyzer` area, like this:

```
analyzer:
  exclude:
    - '**'
  plugins:
    - dart_code_metrics

dart_code_metrics:
  rules:
    - prefer-trailing-comma
```

**Comment by Alex-A4** [comment]
> @Alex-A4 your config has one mistake. Exclude section must be placed in `analyzer` area, like this:
> 
> ```
> analyzer:
>   exclude:
>     - '**'
>   plugins:
>     - dart_code_metrics
> 
> dart_code_metrics:
>   rules:
>     - prefer-trailing-comma
> ```

Oh, thanks!
But what about analyzing of all packages are added with pubspec? Is that a normal behaviour or a bug?

**Comment by dkrutskikh** [comment]
@Alex-A4 can you provide more details?

**Comment by Alex-A4** [comment]
Sure, @dkrutskikh 
1) Add dart_code_metrics to existed project
2) Create `analysis_options.yaml` as said in [usage section](https://pub.dev/packages/dart_code_metrics#usage).
3) Restart IDE to apply plugin. 
Voila, I see a lot of hints related to other packages, added with pubspec. Attaching the image
![image](https://user-images.githubusercontent.com/32712725/115438655-542abf80-a216-11eb-9ed4-fc5ee2fe7025.png)

**Comment by incendial** [comment]
@Alex-A4 isn't it the same problem with `ios/.symlinks` as [described here](https://github.com/dart-code-checker/dart-code-metrics/issues/265#issuecomment-819267637)?

**Comment by Alex-A4** [comment]
@incendial Yeah, this works, thanks! Sorry for duplicate

**Comment by incendial** [comment]
No problem, we probably should consider adding `ios` folder to the default ignores.

**Comment by dkrutskikh** [comment]
@roman-petrov unfortunately, we could not fix your problem in 3.1 release, fix of this problem was moved to release 3.2.

**Comment by roman-petrov** [comment]
@dkrutskikh that's ok. I'm a bit blocked on this issue but I don't have deadline to integrate dart code metrics.
